docs/proposed/conflict-api.rst

```diff
@@ -11,42 +11,77 @@ This document is aimed at programmers working on magic-folder. It is a proposed
 Motivation
 ==========
 
-It is desirable to have an explicit API for dealing with conflicts.
-This allows UI programs to be sure they're accurately indicating intent to magic-folders.
+Under normal operation, magic-folders scans other participants' Snapshots and reflects those changes locally; sometimes, these changes can conflict.
+
+It is desirable to have an explicit API for noticing and resolving conflicts.
+The core of magic-folders operation with users is the filesystem.
+This makes the filesystem also an API and conflicts must be part of that API.
+The command-line interface shall have sub-commands for listing and resolving conficts; these commands will use the HTTP APIs.
+
+
+Filesystem API for Conflicts and Resolution
+===========================================
+
+When a Snapshot is found to conflict with a particluar local file (say ``foo``) a "conflict file" is written beside it, reflecting the other participants' content (``foo.conflict-laptop``).
+That is, ``foo.conflict-laptop`` indicates that the participant "``laptop``" has a conflicting update
+The file ``foo.conflict-laptop`` will contain the downloaded ``.content`` of "``laptop``"'s Snapshot.
+The content of ``foo`` remains what it was when the conflict was detected.
+(Note that when multiple participants exist it's possible to have multiple ``*.confict-*`` files pertaining to a single local file).
+
+List Conflicts
+~~~~~~~~~~~~~~
+
+One can use normal directory-browsing tools such as ``ls`` to notice conflict files.
+
+
+Resolve a Conflict
+~~~~~~~~~~~~~~~~~~
+
+When _all_ ``<relpath>.confict-*`` files for a given root are deleted, the conflict is deemed resolved.
```

Comment:
Somewhere, we should indicate how this gets represented as snapshots. Do we also want to consider how to represent partially resolved conflicts?

There should also be some discussion about how we will notice and accept the resolution of a conflict by another participant, since each participant will have conflict files for every other participant that is in conflict. We probably also need to consider what to do about changes to the conflict files when accepting a conflict resolution (this seems like itself a form of conflict).

Reply:
They are indicated as snapshots with multiple parents.

Reply:
There's no "partially resolved" state, so not sure what you mean by that?

Reply:
What I mean by "partially resolved" is having resolved the conflict with `device-b` (but not `device-a`). It could be nice to record a snapshot that has `device-b` but not `device-a` as a parent. We could (for example) deem that to be what happens when `*.conflict-device-b` is removed.

I'm not sure that is the way we'd want to implement it in the filesystem, and I don't even know if we want to support it. But I think it is reasonable to consider that as a feature we may want, and how it would be implemented. And it is probably worth  considering before we finalize the filesystem API for resolving a conflict.

Reply:
What does that even mean, though, semantically?

Reply:
It would be the same as doing something like say a 2-way merge vs. a n-way merge in git.

Another way of looking at it is that is the same as if we pretended that we hadn't yet noticed a conflict with `device-a`, when resolving the conflict with `device-b`.